### PR TITLE
Add `fft` submodule to `qml.math`

### PR DESCRIPTION
**Pull request by josh146**

### CODE REVIEW

File: pennylane/math/single_dispatch.py [hunk 1]
```diff
@@ -120,6 +120,20 @@ def _take_autograd(tensor, indices, axis=None):
...
+tf_fft_functions = [
+    "fft",
+    "ifft",
+    "fft2",
+    "ifft2",
+]
+
+
+for fn in tf_fft_functions:
+    ar.autoray._SUBMODULE_ALIASES["tensorflow", "fft." + fn] = "tensorflow.signal"
```
Comment (josh146) [inline pennylane/math/single_dispatch.py:132]: This registers the fact that the TensorFlow module is not `tf.fft`, but `tf.signal` :)

File: pennylane/math/single_dispatch.py [hunk 2]
```diff
@@ -120,6 +120,20 @@ def _take_autograd(tensor, indices, axis=None):
...
+
+ar.autoray._FUNC_ALIASES["tensorflow", "fft.fft2"] = "fft2d"
+ar.autoray._FUNC_ALIASES["tensorflow", "fft.ifft2"] = "ifft2d"
```
Comment (josh146) [inline pennylane/math/single_dispatch.py:135]: This now registers that the TF functions have different names

File: pennylane/math/single_dispatch.py [hunk 3]
```diff
@@ -138,6 +152,29 @@ def _take_autograd(tensor, indices, axis=None):
...
+    if a.dtype not in [_i("tf").complex64, _i("tf").complex128]:
+        if a.dtype is _i("tf").float64:
+            a = _i("tf").cast(a, dtype=_i("tf").complex128)
+        else:
+            a = _i("tf").cast(a, dtype=_i("tf").complex64)
+
+    return _i("tf").signal.ifft2d(input=a)
+
+
+ar.register_function("tensorflow", "fft.ifft2", _ifft2_tf)
```
Comment (josh146) [inline pennylane/math/single_dispatch.py:175]: Sometimes, the functionality differs between the different frameworks. Here, I am using `ar.register_function` to register a _custom_ wrapper around the TensorFlow `ifft2`, to account for:

- TensorFlow, unlike Torch and NumPy, only accepts complex tensor input
- TensorFlow does not accept an `axes` or `norm` argument

File: pennylane/math/single_dispatch.py [hunk 4]
```diff
@@ -235,6 +272,15 @@ def _scatter_element_add_tf(tensor, index, value):
 ar.register_function("torch", "shape", lambda x: tuple(x.shape))
 ar.register_function("torch", "gather", lambda x, indices: x[indices])
 
+ar.register_function(
+    "torch",
+    "fft.ifft2",
+    lambda a, s=None, axes=(-2, -1), norm=None: _i("torch").fft.ifft2(
+        input=a, s=s, dim=axes, norm=norm
+    ),
+)
```
Comment (josh146) [inline pennylane/math/single_dispatch.py:281]: The `torch.fft.ifft2` functionality matches almost exactly with the NumPy functionality.

However, it uses a slightly different signature - instead of `a` and `axes`, it uses `input` and `dim`! Here we use `register_function` with a very light lambda function in order to register the different named parameter arguments.

This catches the case where a user might pass keyword arguments, `qml.math.fft.ifft2(tensor, axes=...)`.

File: pennylane/math/single_dispatch.py [hunk 5]
```diff
@@ -138,6 +152,29 @@ def _take_autograd(tensor, indices, axis=None):
...
+            "will always be performed over the inner-most 2 dimensions."
+        )
+
+    if norm is not None:
+        raise ValueError("TensorFlow does not support the 'norm' keyword argument.")
+
+    # TensorFlow only supports FFT of complex tensors
+    if a.dtype not in [_i("tf").complex64, _i("tf").complex128]:
+        if a.dtype is _i("tf").float64:
+            a = _i("tf").cast(a, dtype=_i("tf").complex128)
```
Comment (mariaschuld) [inline pennylane/math/single_dispatch.py:168]: Sure that it is not 64 here and 128 in the next line?

File: pennylane/math/single_dispatch.py [hunk 6]
```diff
@@ -138,6 +152,29 @@ def _take_autograd(tensor, indices, axis=None):
 )
 
 
+def _ifft2_tf(a, s=None, axes=(-2, -1), norm=None):
+    if axes != (-2, -1):
+        raise ValueError(
+            "TensorFlow does not support passing axes; the ifft "
+            "will always be performed over the inner-most 2 dimensions."
+        )
```
Comment (mariaschuld) [inline pennylane/math/single_dispatch.py:160]: Would it be good to check also if `s` is something other than `None`? since it is not used?

**Comment by mariaschuld** [comment]
Nice! I wonder if the `numpy_mimic` is causing the tests to fail?

And what would we need to do, just add tests? Or are the math module functions untested?

**Comment by josh146** [comment]
> Nice! I wonder if the numpy_mimic is causing the tests to fail?

Tests are failing because I stupidly passed `self` to `super()`, I forget that you don't have to and that is automatic 😆 

Just fixed that, so tests should pass now.

> And what would we need to do, just add tests? Or are the math module functions untested?

- Finish ensuring that all TF/Torch FFT functions that are needed are accessible. For now, I imagine just the main ones (`fft`, `ifft`, `fft2` etc.) as well as some of the helper function (`fftshift`) should be sufficient?
- Writing tests 🙂

File: pennylane/math/__init__.py [hunk 7]
```diff
@@ -62,8 +62,22 @@
...
+        return super().__getattribute__(fn)
+
+
+numpy_mimic = NumpyMimic()
+numpy_fft = ar.autoray.NumpyMimic("fft")
+
+
 def __getattr__(name):
-    return getattr(ar.numpy, name)
+    return getattr(numpy_mimic, name)
```
Comment (josh146) [inline pennylane/math/__init__.py:80]: The `NumpyMimic` class that is built into Autoray supports dispatching to the `linalg` and `random` submodules, but _not_ `fft`, so that has been added here.

File: pennylane/math/single_dispatch.py (hunk 5 above)
Comment (josh146) [reply pennylane/math/single_dispatch.py:168]: Oops 🤦

File: pennylane/math/single_dispatch.py (hunk 6 above)
Comment (josh146) [reply pennylane/math/single_dispatch.py:160]: Yep! Definitely.

File: pennylane/math/single_dispatch.py (hunk 5 above)
Comment (josh146) [reply pennylane/math/single_dispatch.py:168]: Oh wait, no, this is what I intended.

- If the input is `float64`, I cast to `complex128` (two double precision floats, one real one imaginary) to retain the same precision.
- Otherwise, I case to `complex64` (two single-precision floats)

File: pennylane/math/single_dispatch.py (hunk 2 above)
Comment (glassnotes) [reply pennylane/math/single_dispatch.py:135]: This is to make the trig function names consistent across all interfaces so you can just so `qml.math.func`, right? Do you mind adding `atan2` to the list? :pray: I ran into this issue earlier in #1384 with `acos` and `atan2` :sweat_smile:

Comment (josh146) [reply pennylane/math/single_dispatch.py:135]: Yep! @glassnotes feel free to add this directly to #1394 --- this PR will likely not be merged for a while, and shouldn't block yours in the meantime 😆 

You will need to add the following two lines:

```python
ar.autoray._SUBMODULE_ALIASES["tensorflow", "arctan2"] = "tensorflow.math"
ar.autoray._FUNC_ALIASES["tensorflow", "arctan2"] = "atan2"
```

The first registers that TF makes the `arctan2` function accessible under the `math` submodule, the second indicates the name change.

(re: `acos` - I can see that one is already registered? What issue were you having?)

Comment (glassnotes) [reply pennylane/math/single_dispatch.py:135]: Ah it ended up being fine and I just needed to add `arctan2` (to both `torch` and `tf`). Thanks! :star2:

**Comment by josh146** [comment]
@glassnotes @mariaschuld @dwierichs... is this still needed? I have completely forgotten the context to this PR :(

**Comment by dwierichs** [comment]
It seems like it is almost done? We have not talked much about differentiability of parameter shift rules or reconstruction functionalities with respect to the frequency or reconstruction/evaluation points since summer. It might be nice to leave this open and merge it in once we want some functionality like this, but for now it does not appear to be needed?

**Comment by josh146** [comment]
@dwierichs yes as far as I can tell, it only needs tests. Okay, let's leave it open for now

**Comment by josh146** [comment]
@dwierichs is this something that is now worth adding?

**Comment by dwierichs** [comment]
I tried to write tests for this.
Can't get the right result in Tensorflow so far, and the following code breaks my  Python instance:
```
import torch
def real_fft(x):
    return torch.real(torch.fft.fft(x))

x = torch.tensor([0.5], requires_grad=True)
torch.autograd.functional.jacobian(real_fft, x)
```

**Comment by josh146** [comment]
@dwierichs did you get the tests working based on the recent commit?

**Comment by dwierichs** [comment]
@josh146 Yes. Complex-valued derivatives are a bit strange in Tensorflow, to me at least. And the torch problem was self-made, I had some local version collision.
This PR should be ready to go soonish, modulo copying the now working tests to the other three registered functions.

**Comment by dwierichs** [comment]
This could use some reviews now :sweat_smile: @josh146

**Comment by josh146** [comment]
Nice work @dwierichs! All looks good to me, and very thorough test coverage. Unfortunately, I cannot approve as I opened this PR 😆 +

**Review by lillian542 — APPROVED** [review]
